src/DelegateRegistry.sol

```diff
@@ -104,199 +100,367 @@ contract DelegateRegistry is IDelegateRegistry {
 
     /// @inheritdoc IDelegateRegistry
-    function delegateForAll(address delegate, bool value, bytes32 data) public override {
-        bytes32 delegationHash = _computeDelegationHashForAll(msg.sender, delegate, data);
-        _setDelegationValues(delegate, delegationHash, value, IDelegateRegistry.DelegationType.ALL, msg.sender, address(0), 0, 0, data);
-        emit IDelegateRegistry.DelegateForAll(msg.sender, delegate, value, data);
+    function delegateForAll(address delegate, bytes32 rights, bool enable) public override {
+        bytes32 hash = _computeDelegationHashForAll(delegate, rights, msg.sender);
+        bytes32 location = _computeDelegationLocation(hash);
+        emit AllDelegated(msg.sender, delegate, rights, enable);
+        if (enable) {
+            if (_loadDelegationBytes32(location, StoragePositions.vault) == "") _pushDelegationHashes(msg.sender, delegate, hash);
+            _writeDelegation(location, StoragePositions.delegate, delegate);
+            _writeDelegation(location, StoragePositions.vault, msg.sender);
+            if (rights != "") _writeDelegation(location, StoragePositions.rights, rights);
+        } else {
+            _writeDelegation(location, StoragePositions.delegate, "");
+            _writeDelegation(location, StoragePositions.vault, 1);
+            if (rights != "") _writeDelegation(location, StoragePositions.rights, "");
+        }
     }
 
     /// @inheritdoc IDelegateRegistry
-    function delegateForContract(address delegate, address contract_, bool value, bytes32 data) public override {
-        bytes32 delegationHash = _computeDelegationHashForContract(msg.sender, delegate, contract_, data);
-        _setDelegationValues(delegate, delegationHash, value, IDelegateRegistry.DelegationType.CONTRACT, msg.sender, contract_, 0, 0, data);
-        emit IDelegateRegistry.DelegateForContract(msg.sender, delegate, contract_, value, data);
+    function delegateForContract(address delegate, address contract_, bytes32 rights, bool enable) public override {
+        bytes32 hash = _computeDelegationHashForContract(contract_, delegate, rights, msg.sender);
+        bytes32 location = _computeDelegationLocation(hash);
+        emit ContractDelegated(msg.sender, delegate, contract_, rights, enable);
+        if (enable) {
+            if (_loadDelegationBytes32(location, StoragePositions.vault) == "") _pushDelegationHashes(msg.sender, delegate, hash);
+            _writeDelegation(location, StoragePositions.contract_, contract_);
+            _writeDelegation(location, StoragePositions.delegate, delegate);
+            _writeDelegation(location, StoragePositions.vault, msg.sender);
+            if (rights != "") _writeDelegation(location, StoragePositions.rights, rights);
+        } else {
+            _writeDelegation(location, StoragePositions.contract_, "");
+            _writeDelegation(location, StoragePositions.delegate, "");
+            _writeDelegation(location, StoragePositions.vault, 1);
```

Comment:
This messes with other getter functions now. On line 333, the `enable` param of getDelegationsFromHashes()` needs to be updated. And we should return a completely empty struct if the delegate doesn't exist